brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+            }
+
+        self.partner_data = partner_data
+        self.best_Q = best_Q
+        return best_Q
+
+    def _upper_tri_sum(self,corr_matrix):
+        """
+        Helper function for traditional method. Given that
+        we want the sum of pairwise correlation, it is essentially the
+        upper triangular of the correlation matrix, without the
+        diagonal. We use a mask to identify that triangle and sum it.
+
+        :param corr_matrix: Spearman correlation matrix.
+        :type corr_matrix: np.ndarray
+        :return: Sum of upper triangular, less the diagonal.
+        :rtype: float
+        """
+        m = corr_matrix.shape[0]
+        r = np.arange(m)
+        mask = r[:,None] < r
+
+        return np.sum(corr_matrix[mask])
+
+    def _extended_method(self, estimator: int) -> list:
+        """
+        The extended method identifies the best partners based on spearman multivariate correlation.
+        This function serves as a helper, pointing to the specific implementation of the extended
+        method that the user would like to use.
+
+        :param estimator: (int): The type of estimator that you would like to use (p1,p2,p3)
+        :raises Exception: In the event an invalid estimator is chosen, error will be raised.
+        :return: The best partners based on multivariate spearman correlation will be returned.
+        :rtype: list
+        """
+        # d, number of potential partners, inclusive of the target ticker.
+        d = self.no_of_partners + 1
+
+        # Term required for computation of estimators.
+        h_d = (d + 1) / (2 ** d - d - 1)
+
+        # Setting it to private attributes.
+        self.__d = d
+        self.__h_d = h_d
+
+        if estimator == 1:
+            best_Q = self._extended_method_estimator_1()
+        elif estimator == 2:
+            best_Q = self._extended_method_estimator_2()
+        elif estimator == 3:
+            best_Q = self._extended_method_estimator_3()
+        else:
+            raise Exception('Invalid estimator provided. See documentation for estimator 1, 2 and 3.')
+
+        return best_Q
+
+    def _extended_method_estimator_1(self) -> list:
+        """
+        This function loops through all potential partners, using estimator 1
+        provided by the paper to estimate multivariate correlation.
+
+        :return: The best partners based on multivariate spearman correlation (estimator 1) will be returned.
+        :rtype: list
+
+        https://econpapers.repec.org/article/eeestapro/v_3a77_3ay_3a2007_3ai_3a4_3ap_3a407-416.htm
+        """
+
+        highest_p_1 = 0
+
+        # Initialise the dictionary to collect data on all
+        # possible groupings for descriptive statistics.
+        partner_data = {}
+
+        for i in tqdm(range(len(self.all_possible_partners))):
+
+            # Converting to list as itertools combination generates an itertools object.
```

Comment:
I'm seeing a lot of duplicate code here. As noted in your write-up, this should be more modular.

Reply:
Never have duplicate code.